Gems/AtomLyIntegration/CommonFeatures/Code/Source/PostProcess/EditorModeFeedback/EditorEditorModeFeedbackSystemComponent.cpp

```diff
@@ -0,0 +1,268 @@
+        {
+            AZStd::vector<RPI::MeshDrawPacket> meshDrawPackets;
+            const RPI::Model& model = *RPI::Model::FindOrCreate(modelAsset);
+            const auto modelLodIndex = RPI::ModelLodUtils::SelectLod(view, worldTM, model);
+            const Data::Asset<RPI::ModelLodAsset>& modelLodAsset = modelAsset->GetLodAssets()[0];
+            RPI::ModelLod& modelLod = *RPI::ModelLod::FindOrCreate(modelLodAsset, modelAsset).get();
+        
+            for (auto i = 0; i < modelLod.GetMeshes().size(); i++)
+            {
+                RPI::MeshDrawPacket drawPacket(modelLod, i, material, meshObjectSrg);
+                meshDrawPackets.push_back(AZStd::move(drawPacket));
+            }
+        
+            return meshDrawPackets;
+        }
+
+        //! Creates the mask shader resource group for the specified mesh.
+        static Data::Instance<RPI::ShaderResourceGroup> CreateMaskShaderResourceGroup(
+            const MeshFeatureProcessorInterface* featureProcessor,
+            const MeshFeatureProcessorInterface::MeshHandle& meshHandle,
+            Data::Instance<RPI::Material> maskMaterial)
+        {
+            const auto& shaderAsset = maskMaterial->GetAsset()->GetMaterialTypeAsset()->GetShaderAssetForObjectSrg();
+            const auto& objectSrgLayout = maskMaterial->GetAsset()->GetObjectSrgLayout();
+            const auto maskMeshObjectSrg = RPI::ShaderResourceGroup::Create(shaderAsset, objectSrgLayout->GetName());
+
+            // Set the object id so the correct MVP matrices can be selected in the shader
+            const auto objectId = featureProcessor->GetObjectId(meshHandle).GetIndex();
+            RHI::ShaderInputNameIndex objectIdIndex = "m_objectId";
+            maskMeshObjectSrg->SetConstant(objectIdIndex, objectId);
+
+            //! Set the id to write to the entity mask texture.
+            RHI::ShaderInputNameIndex maskIdIndex = "m_maskId";
+            maskMeshObjectSrg->SetConstant(maskIdIndex, 1);
+            
+            maskMeshObjectSrg->Compile();
+
+            return maskMeshObjectSrg;
+        }
+
+        //! Gets the view for the specified scene.
+        static const RPI::ViewPtr GetViewFromScene(const RPI::Scene* scene)
+        {
+            const auto viewportContextRequests = RPI::ViewportContextRequests::Get();
+            const auto viewportContext = viewportContextRequests->GetViewportContextByScene(scene);
+            const RPI::ViewPtr viewPtr = viewportContext->GetDefaultView();
+            return viewPtr;
+        }
+
+        //! Creates the material for the mask pass shader.
+        static Data::Instance<RPI::Material> CreateMaskMaterial()
+        {
+            const AZStd::string path = "shaders/postprocessing/editormodemask.azmaterial";
+            const auto materialAsset = GetAssetFromPath<RPI::MaterialAsset>(path, Data::AssetLoadBehavior::PreLoad, true);
+            const auto maskMaterial = RPI::Material::FindOrCreate(materialAsset);
```

Comment:
FindOrCreate is going to do a blocking load if the asset isn't already loaded. Blocking loads can both cause a hitch and have a tendency to lead to deadlocks. It's probably worth keeping a Data::Asset<RPI::MaterialAsset> m_maskMaterialAsset member variable, doing a non-blocking queue load early on (like during Activate) so that by the time you get here the asset is likely already loaded.

Reply:
+1 on that - can lead to deadlock.
The alternative is queue load and listen to OnAssetReady or alike as Tommy hinted - it is much safer and should be the preferred method if can be implemented without over complicating your logic.

Reply:
IIRC i did try getting the mask material on `Activate` but because it's currently a system component it gets activated before the Atom gem is up and running :( When we move everything over to AtomLyIntegration i think this specific issue will be resolved easily enough 👍

Reply:
There are a couple of options you can explore when you move to AtomLyIntegration, I'm not sure what works best. But you could make your system component dependent on an Atom system component so that it activates after the Atom system components. I don't know if that's enough for loading a material asset to be functional or not. If not, we also have an AzRender::Boostrap::NotificationBus and probably some other bus you could listen to to find out when you're ready to load a material asset. You don't necessarily have to be ready to render, or ready to create the instance using FindOrCreate. Just load the asset so that you don't have a blocking asset load when you finally do want to call FindOrCreate.